src/CacheManager.Core/BaseCacheManager.Keys.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using CacheManager.Core.Logging;
+
+namespace CacheManager.Core
+{
+    public sealed partial class BaseCacheManager<TCacheValue>
+    {
+        /// <inheritdoc />
+        override public IEnumerable<string> Keys(string pattern, string region)
+        {
+            CheckDisposed();
+
+            var keys = new HashSet<string>();
+
+            if (_logTrace)
+            {
+                Logger.LogTrace("Keys [{0}:{1}] started.", region, pattern);
+            }
+
+            for (var handleIndex = 0; handleIndex < _cacheHandles.Length; handleIndex++)
+            {
+                var handle = _cacheHandles[handleIndex];
+                var handleKeys = handle.Keys(pattern ?? "*", region);
+                keys.UnionWith(handleKeys);
+            }
+
+            if (_logTrace)
+            {
+                Logger.LogTrace("Keys [{0}:{1}] found {2} keys.", region, pattern, keys.Count);
+            }
+
+            if (region != null)
+            {
+                var after = region.Length + 1; // +1 for the :
+                return keys.Select(k => k.Substring(after));
```

Comment:
Addressed in [3ccc673](https://github.com/MichaCo/CacheManager/pull/161/commits/3ccc6738a0a41104d40fffe6d8020e5f8c682331)

I moved the prefix logic down into the handlers.